dialogue-jmh/src/main/java/com/palantir/dialogue/core/EndToEndBenchmark.java

```diff
@@ -0,0 +1,145 @@
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+@State(Scope.Benchmark)
+@Warmup(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 3, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@Fork(value = 1)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@BenchmarkMode(Mode.Throughput)
+@SuppressWarnings({"VisibilityModifier", "DesignForExtension"})
+public class EndToEndBenchmark {
+
+    private Undertow undertow;
+    private ExecutorService blockingExecutor;
+
+    private SampleServiceBlocking blocking;
+    private SampleServiceAsync async;
+
+    @Setup
+    public void before() {
+        undertow = Undertow.builder()
+                // .addHttpListener(0, "localhost", new BlockingHandler(exchange -> exchange.setStatusCode(200)))
+                .addHttpListener(0, "localhost", new ResponseCodeHandler(200))
+                .build();
+        undertow.start();
+
+        TaggedMetricRegistry metrics = new DefaultTaggedMetricRegistry();
+
+        blockingExecutor = Tracers.wrap(
+                "dialogue-blocking-channel",
+                Executors.newCachedThreadPool(MetricRegistries.instrument(
+                        metrics,
+                        new ThreadFactoryBuilder()
+                                .setNameFormat("dialogue-blocking-channel-%d")
+                                .setDaemon(true)
+                                .build(),
+                        "dialogue-blocking-channel")));
+
+        DialogueClients.ReloadingFactory clients = DialogueClients.create(Refreshable.only(null))
+                .withUserAgent(TestConfigurations.AGENT)
+                .withTaggedMetrics(metrics)
+                .withBlockingExecutor(blockingExecutor);
+
+        ServiceConfiguration serviceConf = ServiceConfiguration.builder()
+                .addUris(getUri(undertow))
+                .security(TestConfigurations.SSL_CONFIG)
+                .build();
+
+        blocking = clients.getNonReloading(SampleServiceBlocking.class, serviceConf);
+        async = clients.getNonReloading(SampleServiceAsync.class, serviceConf);
+    }
+
+    @TearDown
+    public void after() {
+        undertow.stop();
+        MoreExecutors.shutdownAndAwaitTermination(RetryingChannel.sharedScheduler.get(), 1, TimeUnit.SECONDS);
+        MoreExecutors.shutdownAndAwaitTermination(blockingExecutor, 1, TimeUnit.SECONDS);
+    }
+
+    @Threads(1)
```

Comment:
We can specify a sane number of threads everywhere and run with `-t 1` for other counts to avoid duplication.